stringlength.md

```diff
@@ -0,0 +1,112 @@
+# Proposal to add Suggested Lengths for String Signals in the VSS
+
+This document covers a concept proposal to add to the VSS .vspec files an additional keyword 
+(`stringsize:` or similar) to assist the code generators (in vss-tools) in setting an appropriate
+default size for strings.
+
+This new keyword would be added to the .vspec files for all `datatype: string` elements (approx. 85 places),
+and would also affect the arrays-of-strings (`datatype: string[]`, approx 6 places) to set a
+correctly-sized default boundary length on these string members.
+
+A few additional questions have been added below as well, for discussion in VSS meetings.
+
+
+## Rationale
+
+When creating software for a system with safety, security, or resource constraints, the use of 
+unbounded variables is strongly discouraged.   Unless extra steps are taken to limit their size,
+they can act as a vector for instability and exploitation.
+
+In practice, the signals in the VSS will never require unbounded lengths.  Some string signals
+in the VSS include the length in its `description:` field, but this information is not available for 
+every `string` signal.  Specifying the `stringlength`, and the `arraysize` of VSS elements can help
+create more concise implementations of the VSS, and reduce the burden on users.
+
+
+## Proposed Implementation
+
+The proposed implementation would add in the .vspec files a new key:value pair (herein called `stringlength:`) 
+to set the string length of any signal with a `string` or `string[]` datatype.  
+When used with arrays-of-strings (`string[]`), this keyword sets the max length of the strings in the array, while
+the existing `arraysize` keyword sets the number of strings in the array.
+
+
+## Examples from VSS
+
+An example implementation might look like:
+```
+VehicleIdentification.VIN:
+  datatype: string
+  stringlength: 17

```

Comment:
Strings (or collections) have two properties:
1. capacity - Maximum allowed size
2. length - Occupied size

Taking the example of C++ basic string, these are modeled as 
max_size - https://en.cppreference.com/w/cpp/string/basic_string/max_size
size/length - https://en.cppreference.com/w/cpp/string/basic_string/size

Size/length seem to be implementation detail as they may vary at the data sample level.

I assume what we are trying to model here is the capacity/max size. Is that correct?

Reply:
I would see it as well like that. But  maybe @neil-rti  can comment?